src/lib/pubkey/hss_lms/hss.h

```diff
@@ -0,0 +1,402 @@
+
+      /**
+       * @brief The object identifier for HSS-LMS
+       */
+      OID object_identifier() const;
+
+      /**
+       * @brief The algorithm name for HSS-LMS
+       */
+      std::string algo_name() const { return "HSS-LMS"; }
+
+      /**
+       * @brief Verify a HSS-LMS signature.
+       *
+       * See RFC 8554 6.3.
+       *
+       * @param msg The signed message.
+       * @param sig The already parsed HSS-LMS signature.
+       * @return True iff the signature is valid.
+       */
+      bool verify_signature(std::span<const uint8_t> msg, const HSS_Signature& sig) const;
+
+   private:
+      /**
+       * @brief Returns the number of layers of LMS trees in the HSS tree.
+       */
+      HSS_Level L() const { return m_L; }
+
+      HSS_Level m_L;
+      LMS_PublicKey m_top_lms_pub_key;
+};
+
+/**
+ * @brief A HSS-LMS signature.
+ *
+ * Format according to RFC 8554:
+ * u32str(Nspk) || sig[0] || pub[1] || ... || sig[Nspk-1] || pub[Nspk] || sig[Nspk]
+ */
+class BOTAN_TEST_API HSS_Signature final {
+   public:
+      /**
+       * @brief A LMS public key signed by the HSS layer above it.
+       *
+       * signed_pub_key[i] = sig[i] || pub[i+1],
+       * for i between 0 and Nspk-1, inclusive.
+       */
+      class Signed_Pub_Key {
+         public:
+            /**
+             * @brief Constructor for a new sig-pubkey-pair
+             */
+            Signed_Pub_Key(LMS_Signature sig, LMS_PublicKey pub);
+
+            /**
+             * @brief The signature of the public key
+             */
+            const LMS_Signature& signature() const { return m_sig; }
+
+            /**
+             * @brief The signed public key
+             */
+            const LMS_PublicKey& public_key() const { return m_pub; }
+
+         private:
+            LMS_Signature m_sig;
+            LMS_PublicKey m_pub;
+      };
+
+      /**
+       * @brief Parse a HSS-LMS signature.
+       *
+       * @param sig_bytes The signature bytes to parse.
+       * @return The parsed HSS-LMS signature.
+       * @throws Decoding_Error If parsing the signature fails.
+       */
+      static HSS_Signature from_bytes_or_throw(std::span<const uint8_t> sig_bytes);
+
+      /**
+       * @brief Returns the size a signature would have in its encoded format.
+       *
+       * @param params The HSS-LMS parameters.
+       * @return size_t The expected size in bytes.
+       */
+      static size_t size(const HSS_LMS_Params& params);
+
+      /**
+       * @brief Returns the number of signed public keys (Nspk = L-1).
+       */
+      uint8_t Nspk() const { return m_Nspk; }
+
+      /**
+       * @brief Returns the signed LMS key signed by a specific layer.
+       *
+       * @param layer The layer by which the LMS key is signed.
+       * @return The LMS key and the signature by its parent layer.
+       */
+      const Signed_Pub_Key& signed_pub_key(uint8_t layer) const { return m_signed_pub_keys.at(layer); }
```

Comment:
This smells like the already existing strong type HSS_Level. I don't know why we didn't use this one.

Reply:
> This smells like the already existing strong type HSS_Level. I don't know why we didn't use this one.

Not using the same strong type here was a deliberate choice by me, and I still would prefer to not use the same strong types for `Nspk` and `L`.


The definitions of `L` and `Nspk` according to the RFC:
> The number of levels is denoted as L and is between one and eight, inclusive.

> Nspk = L-1 denotes the number of signed public keys.

Levels are indexing the different layers of the HSS tree,
the range is `[0,L)`, 0 being the root.
Note that the RFC nether uses the term layer, unsure why we started to use it.

For me the subtle differences were very confusing at the beginning, and so I wanted to especially use different types for `L` and `Nspk` to make it very clear that they can not just be interchanged with each other. Although they are of course very closely related to each other.

Now do we really want to have three different strong types for this, one each for level, `Nspk` and `L`?
I'm unsure but probably that is a little overkill.
But again we should at least avoid using the same strong type for both `Nspk` and `L`, as otherwise the point of using strong types to avoid misuse is kind of defeated.